### ISSUE TITLE
Scrape mehod always returns empty array

### ISSUE THREAD

**Issue by alperenbaskaya58** [pevers/images-scraper#113]
I was using this package to grab image links. Even the sample code is not working returns empty array. What could be the problem?

`var Scraper = require('images-scraper');

const google = new Scraper({
  puppeteer: {
    headless: false,
  },
  //safe : false,
  //userAgent: 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko)', // the user agent

});

(async () => {
    
  const results = await google.scrape('banana', 5);
  console.log('results', results);
})();`

**Comment by kanjieater** [comment]
@pevers Any way you could update this for us?

**Comment by pevers** [comment]
Hey @kanjieater , I haven't looked into it yet as I was busy with other things. But I'll try to have a look soon. In the meantime, if there is anyone with a solution I would be glad to merge it!

**Comment by gojodev** [comment]
This happened to me as well, so instead I just used puppeteer from node

This is what I _currently_ have:

```js
import puppeteer from 'puppeteer';


function imgScrape(queries) {
    (async () => {
        try {
            const browser = await puppeteer.launch({ headless: true });
            const page = await browser.newPage();

            for (const query of queries) {
                await page.goto(`https://www.google.com/search?tbm=isch&q=${query}`);

                // Scroll to the bottom of the page to load more images
                await page.evaluate(async () => {
                    for (let i = 0; i < 10; i++) {
                        window.scrollBy(0, window.innerHeight);
                        await new Promise(resolve => setTimeout(resolve, 500)); // Wait for more images to load
                    }
                });

                // Wait for images to be loaded
                await page.waitForSelector('img');

                // Extract image URLs
                const images = await page.evaluate(() => {
                    const imageElements = document.querySelectorAll('img');
                    const urls = [];
                    imageElements.forEach(img => {
                        const url = img.src;
                        if (url.startsWith('http') && !url.includes('google')) {
                            urls.push(url);
                        }
                    });
                    return urls.slice(0, 3); // Limit to first 3 image URLs
                });

                console.log(`\nResults for query: ${query}`);
                images.forEach((url, index) => {
                    console.log(`Image ${index + 1}: ${url}`);
                });
            }

            await browser.close();

        } catch (err) {
            console.error('An error occurred:', err);
        }
    })();
}

imgScrape(['python programming language logo']) // Fill array as needed
```

**TLDR; don't use "images-scraper" and use "puppeteer" **

**Comment by kanjieater** [comment]
> This happened to me as well, so instead I just used puppeteer from node
> 
> This is what I _currently_ have:
> 
> ```js
> import puppeteer from 'puppeteer';
> 
> async function imgScrape(queries) {
>     try {
>         const browser = await puppeteer.launch({ headless: true });
>         const page = await browser.newPage();
>         var images;
>         for (const query of queries) {
>             await page.goto(`https://www.google.com/search?tbm=isch&q=${query}`);
> 
>             // Scroll to the bottom of the page to load more images
>             await page.evaluate(async () => {
>                 for (let i = 0; i < 10; i++) {
>                     window.scrollBy(0, window.innerHeight);
>                     await new Promise(resolve => setTimeout(resolve, 500)); // Wait for more images to load
>                 }
>             });
> 
>             // Wait for images to be loaded
>             await page.waitForSelector('img');
> 
>             // Extract image URLs
>             images = await page.evaluate(() => {
>                 const imageElements = document.querySelectorAll('img');
>                 const urls = [];
>                 imageElements.forEach(img => {
>                     const url = img.src;
>                     if (url.startsWith('http') && !url.includes('google')) {
>                         urls.push(url);
>                     }
>                 });
>                 return urls.slice(0, 3); // Limit to first 3 image URLs
>             });
>         }
> 
>         await browser.close();
>         return images;
> 
>     } catch (err) {
>         console.error('An error occurred:', err);
>     }
> }
> 
> const urls = Promise.resolve(imgScrape(['python programming language logo']));
> 
> urls.then((urls) => {
>     console.log(urls);
> })
> ```
> 
> **TLDR; don't use "images-scraper" and use "puppeteer"**

I tried this in headless mode today, and unfortunately could never get it to return, it just hung indefinitely for me.

**Comment by rohitnairtech** [comment]
Here's an improved version from the one posted by @gojodev 


```
import puppeteer from 'puppeteer';

async function imgScrape(queries, limit) {
    try {
        const browser = await puppeteer.launch({ headless: true });
        const page = await browser.newPage();
        const allImages = new Set();
        for (const query of queries) {
            await page.goto(`https://www.google.com/search?tbm=isch&q=${query}`);
            
            await page.evaluate(async () => {
                for (let i = 0; i < 10; i++) {
                    window.scrollBy(0, window.innerHeight);
                    await new Promise(resolve => setTimeout(resolve, 500));
                }
            });

            await page.waitForSelector('img');

            const images = await page.evaluate(() => {
                const imageElements = document.querySelectorAll('img');
                const urls = [];
                imageElements.forEach(img => {
                    const url = img.src;
                    if (url.startsWith('http') && !url.includes('google')) {
                        urls.push(url);
                    }
                });
                return urls;
            });

            images.forEach(url => allImages.add(url));
        }

        await browser.close();

        return Array.from(allImages).slice(0, limit);

    } catch (err) {
        console.error('An error occurred:', err);
    }
}

imgScrape(['cat'], 5).then((urls) => {
    console.log(urls);
});
```

Key changes

1. Using `set` for unique
2. Added limit param, converted `set` to `array`
3. Direct Promise Handling - instead of using Promise.resolve over an existing promise.

**Comment by rohitnairtech** [comment]
Hello guys, I have created a PR for `image-scraper:master` at https://github.com/pevers/images-scraper/pull/114/files

And a working fork is at https://github.com/rohitnairtech/images-scraper

Key changes

1. Added support for download
2. Shuffle between userAgents to avoid detection
3. Added default puppeteer configs
4. Renamed `scraper.js` inside of google to `index.js` (dont want to write scrapers in the import *lazy* :) ) 
5. Refactored winston logger
6. Added axios package :package: for downloading
7. Added fs and path package to create directories and files in sync
8. Removed `_scrapePage` `_scrape` `_isScrollable` `_isButtonVisible` `_clickAllImages`
9. Added two new methods - `downloadImages` `getImageUrl`

And most importantly, the code works! :tada:

**Comment by rohitnairtech** [comment]
@pevers let me know if we can contribute on this project or create a new one? (wouldn't want to create a new project - there are so many of them already)
At the moment I have a use-case for this package and need it working and with features.

Secondly, I noticed that the latest one published on NPM is not the same as the latest `master` here.

@TrifiAmanallah @kanjieater @gojodev please free feel to check https://github.com/rohitnairtech/images-scraper and give your suggestion :) 

Happy to code!

**Comment by pevers** [comment]
> @pevers let me know if we can contribute on this project or create a new one? (wouldn't want to create a new project - there are so many of them already) At the moment I have a use-case for this package and need it working and with features.
> 
> Secondly, I noticed that the latest one published on NPM is not the same as the latest `master` here.
> 
> @TrifiAmanallah @kanjieater @gojodev please free feel to check https://github.com/rohitnairtech/images-scraper and give your suggestion :)
> 
> Happy to code! 👨‍💻

Hey @rohitnairtech , thank you for making the fix! I'll review your PR.

**Comment by pevers** [comment]
@rohitnairtech I think there is still an issue with the cookie header. Which is probably regional. So it still seems a bit flaky but I will merge this for now since it is already a huge improvement

**Comment by rohitnairtech** [comment]
@pevers thank you Peter! I'm open to contributing more, in case there are any more features/improvements that the community would like. This is literally the first package that shows up when you search `google image scrapper nodejs`. :)

**Comment by tmhglnd** [comment]
> [@rohitnairtech](https://github.com/rohitnairtech) I think there is still an issue with the cookie header. Which is probably regional. So it still seems a bit flaky but I will merge this for now since it is already a huge improvement

Yes, I am encountering the issue with the cookie header (I'm in NL). My work-around for now is to stop using Google for the image search and switch to Duckduckgo. I used the code from @rohitnairtech above, but instead of ```await page.goto(`https://www.google.com/search?tbm=isch&q=${query}`);``` I used: 

```js
await page.goto(`https://duckduckgo.com/?q=${query}&atb=v314-1&iar=images&iax=images&ia=images`);
```

Just wanted to share this if others are encountering the same issue.
